File: bench/report.js
#!/usr/bin/env node
/**
 * Generate Markdown report from benchmark JSONL result files.
 * Each .jsonl file contains one JSON line per binary run (Catch2 XML → xml-js compact JSON).
 * Usage: node report.js <file.jsonl> [...]
 */

const fs = require('fs');
const os = require('os');
const { execSync } = require('child_process');

// Format large numbers (requests/sec)
function formatReqPerSec(reqPerSec) {
    if (reqPerSec >= 1000000) {
        return (reqPerSec / 1000000).toFixed(2) + 'M';
    } else if (reqPerSec >= 1000) {
        return (reqPerSec / 1000).toFixed(2) + 'K';
    }
    return reqPerSec.toFixed(0);
}

// Format data size
function formatDataSize(bytes) {
    if (bytes >= 1024) {
        return (bytes / 1024).toFixed(2) + ' KB';
    }
    return bytes + ' bytes';
}

// Get system information
function getSystemInfo() {
    const info = {
        platform: os.type(),
        arch: os.machine() || os.arch(),
        osVersion: '',
        cpu: '',
        cpuCores: os.cpus().length,
        memory: '',
        compiler: ''
    };

    const memBytes = os.totalmem();
    info.memory = `${Math.round(memBytes / (1024 * 1024 * 1024))} GB`;

    try {
        if (process.platform === 'darwin') {
            info.osVersion = execSync('sw_vers -productVersion', { encoding: 'utf-8' }).trim();
            const osBuild = execSync('sw_vers -buildVersion', { encoding: 'utf-8' }).trim();
            info.osVersion += ` (${osBuild})`;
            info.cpu = execSync('sysctl -n machdep.cpu.brand_string', { encoding: 'utf-8' }).trim();
            try {
                info.compiler = execSync('clang --version 2>/dev/null | head -1', { encoding: 'utf-8', shell: '/bin/bash' }).trim();
            } catch (e) { info.compiler = 'Unknown'; }
        } else if (process.platform === 'linux') {
            try {
                const osRelease = fs.readFileSync('/etc/os-release', 'utf-8');
                const nameMatch = osRelease.match(/PRETTY_NAME="([^"]+)"/);
                info.osVersion = nameMatch ? nameMatch[1] : 'Linux';
            } catch (e) { info.osVersion = 'Linux'; }
            try {
                const cpuInfo = fs.readFileSync('/proc/cpuinfo', 'utf-8');
                const m = cpuInfo.match(/model name\s*:\s*(.+)/);
                if (m) info.cpu = m[1].trim();
            } catch (e) { }
            try {
                info.compiler = execSync('gcc --version 2>/dev/null | head -1', { encoding: 'utf-8', shell: '/bin/bash' }).trim();
            } catch (e) {
                try {
                    info.compiler = execSync('clang --version 2>/dev/null | head -1', { encoding: 'utf-8', shell: '/bin/bash' }).trim();
                } catch (e2) { info.compiler = 'Unknown'; }
            }
        } else {
            info.osVersion = os.release();
            info.cpu = os.cpus()[0]?.model || 'Unknown';
            info.compiler = 'Unknown';
        }
    } catch (e) {
        info.osVersion = os.release();
        info.cpu = os.cpus()[0]?.model || 'Unknown';
        info.compiler = 'Unknown';
    }

    return info;
}

// Category metadata (keyed by TestCase name)
const CATEGORY_META = {
    'Header Count': {
        description: 'Measures parsing time scaling with increasing header counts. hwire `(LC)` variants include lowercase key conversion.'
    },
    'Header Value Length': {
        description: 'Measures how parsing time scales with header value size. hwire `(LC)` variants include lowercase key conversion.'
    },
    'Case Sensitivity': {
        description: 'Measures header name normalization cost (lowercase vs mixed case). hwire `(LC)` variants include lowercase key conversion.'
    },
    'Real-World Requests': {
        description: 'Typical requests from browsers, REST APIs, and mobile apps. hwire `(LC)` variants include lowercase key conversion.'
    },
    'Baseline': {
        description: 'Minimum parsing cost for comparison. hwire `(LC)` variants include lowercase key conversion.'
    },
    'Real-World Responses': {
        description: 'Typical responses from web servers and CDNs. hwire `(LC)` variants include lowercase key conversion.'
    }
};

// Display order of categories in report
const CATEGORY_ORDER = [
    'Header Count',
    'Header Value Length',
    'Case Sensitivity',
    'Real-World Requests',
    'Baseline',
    'Real-World Responses'
];

// Format time for table (shorter format)
function formatTimeShort(ns) {
    if (ns >= 1000000) {
        return (ns / 1000000).toFixed(2) + ' ms';
    } else if (ns >= 1000) {
        return (ns / 1000).toFixed(2) + ' μs';
    } else {
        return ns.toFixed(2) + ' ns';
    }
}

// Parse variant name to display name
function getVariantDisplayName(variant) {
    switch (variant) {
        case 'nosimd': return 'No-SIMD';
        case 'sse2': return 'SSE2';
        case 'sse42': return 'SSE4.2';
        case 'avx2': return 'AVX2';
        case 'neon': return 'NEON';
        case 'simd': return 'SIMD';
        default: return variant.toUpperCase();
    }
}

// Extract parser and variant from binary name
// bench_hwire_sse42 -> { parser: 'hwire', variant: 'sse42' }
// bench_hwire_resp_sse42 -> { parser: 'hwire', variant: 'sse42' }
// bench_httparse_simd -> { parser: 'httparse', variant: 'simd' }
function extractParserVariant(exeName) {
    const parts = exeName.split('_');
    // parts[0] = 'bench', parts[1] = parser
    if (parts.length < 3) return { parser: parts[1] || exeName, variant: 'default' };
    // bench_X_resp_Y -> parser=X, variant=Y
    if (parts[2] === 'resp') {
        return { parser: parts[1], variant: parts[3] || 'default' };
    }
    // bench_X_Y -> parser=X, variant=Y
    // bench_X_Y_Z -> parser=X, variant=Y_Z (e.g. future multi-part variant)
    return { parser: parts[1], variant: parts.slice(2).join('_') };
}

// Parse all .jsonl files and return flat results array
function parseJsonlFiles(files) {
    const results = [];

    for (const filepath of files) {
        const lines = fs.readFileSync(filepath, 'utf-8').split('\n').filter(l => l.trim());
        for (const line of lines) {
            let run;
            try {
                run = JSON.parse(line);
            } catch (e) {
                console.warn(`Warning: failed to parse JSON line in ${filepath}: ${e.message}`);
                continue;
            }

            if (!run.runName || !run.testCases) continue;

            const { parser, variant } = extractParserVariant(run.runName);
            const variantDisplay = getVariantDisplayName(variant);

            for (const tc of run.testCases) {
                // tc.name = "Category, Data Label" (e.g. "Header Count, 8 Headers")
                const tcParts = tc.name.split(', ');
                const category = tcParts[0] || 'Unknown';
                const dataLabel = tcParts[1] || '';

                for (const bm of tc.benchmarks || []) {
                    // bm.name = "433 B" or "433 B, LC"
                    const bmParts = bm.name.split(', ');
                    const sizeStr = bmParts[0];
                    const opts = bmParts.slice(1);
                    const effectiveVariantDisplay = opts.length > 0
                        ? `${variantDisplay} (${opts.join(', ')})`
                        : variantDisplay;

                    // benchName = "Data Label, Size" keeps the same table display
                    const benchName = dataLabel ? `${dataLabel}, ${sizeStr}` : sizeStr;

                    const sizeMatch = sizeStr.match(/^(\d+)\s*B$/);
                    const dataSize = sizeMatch ? parseInt(sizeMatch[1], 10) : 0;

                    results.push({
                        category,
                        benchName,
                        parser,
                        variant,
                        variantDisplay: effectiveVariantDisplay,
                        meanNs: bm.meanNs,
                        lowMeanNs: bm.lowMeanNs,
                        highMeanNs: bm.highMeanNs,
                        stdDevNs: bm.stdDevNs,
                        dataSize
                    });
                }
            }
        }
    }

    return results;
}

// Group results by category -> benchName -> [entries]
function groupResults(results) {
    const groups = {};
    for (const r of results) {
        if (!groups[r.category]) groups[r.category] = {};
        if (!groups[r.category][r.benchName]) groups[r.category][r.benchName] = [];
        groups[r.category][r.benchName].push(r);
    }
    return groups;
}

// Generate Markdown table for one benchmark
function generateTable(entries, benchName) {
    entries.sort((a, b) => a.meanNs - b.meanNs);
    const dataSize = entries[0]?.dataSize || 0;

    const maxParserLen = Math.max(8, ...entries.map(r => r.parser.length));
    const maxVariantLen = Math.max(7, ...entries.map(r => r.variantDisplay.length));

    let table = `### ${benchName}\n\n`;
    if (dataSize > 0) {
        table += `Data Size: ${formatDataSize(dataSize)}\n\n`;
    }

    table += '| ' + 'Parser'.padEnd(maxParserLen) +
        ' | ' + 'Variant'.padEnd(maxVariantLen) +
        ' | ' + 'Mean'.padStart(10) +
        ' | ' + 'Std Dev'.padStart(9) +
        ' | ' + '95% CI'.padEnd(21) +
        ' | ' + 'Req/sec'.padStart(8) +
        ' | ' + 'MB/sec'.padStart(7) +
        ' |\n';

    table += '| ' + '-'.repeat(maxParserLen) +
        ' | ' + '-'.repeat(maxVariantLen) +
        ' | ' + '-'.repeat(10) +
        ' | ' + '-'.repeat(9) +
        ' | ' + '-'.repeat(21) +
        ' | ' + '-'.repeat(8) +
        ' | ' + '-'.repeat(7) +
        ' |\n';

    for (const r of entries) {
        const meanStr = formatTimeShort(r.meanNs);
        const stdDevStr = formatTimeShort(r.stdDevNs);
        const ciStr = formatTimeShort(r.lowMeanNs) + ' - ' + formatTimeShort(r.highMeanNs);
        const reqPerSec = 1_000_000_000 / r.meanNs;
        const mbPerSec = dataSize > 0 ? (dataSize * reqPerSec) / 1_000_000 : 0;

        table += '| ' + r.parser.padEnd(maxParserLen) +
            ' | ' + r.variantDisplay.padEnd(maxVariantLen) +
            ' | ' + meanStr.padStart(10) +
            ' | ' + stdDevStr.padStart(9) +
            ' | ' + ciStr.padEnd(21) +
            ' | ' + formatReqPerSec(reqPerSec).padStart(8) +
            ' | ' + (dataSize > 0 ? mbPerSec.toFixed(2) : '-').padStart(7) +
            ' |\n';
    }

    return table;
}

// Generate ASCII bar chart
function generateBarChart(entries, maxBarLength = 30) {
    entries.sort((a, b) => a.meanNs - b.meanNs);
    const maxNs = Math.max(...entries.map(r => r.meanNs));
    const baselineNs = entries[0].meanNs;
    const maxLabelLen = Math.max(...entries.map(r => (r.parser + ' ' + r.variantDisplay).length));
    const lines = [];

    for (let i = 0; i < entries.length; i++) {
        const r = entries[i];
        const label = (r.parser + ' ' + r.variantDisplay).padEnd(maxLabelLen);
        const barLength = Math.round((r.meanNs / maxNs) * maxBarLength);
        const bar = '█'.repeat(barLength);
        const timeStr = r.meanNs >= 1000
            ? (r.meanNs / 1000).toFixed(1) + ' μs'
            : r.meanNs.toFixed(1) + ' ns';
        const annotation = i === 0
            ? '(baseline)'
            : `(${((r.meanNs - baselineNs) / baselineNs * 100).toFixed(1)}% slower)`;
        lines.push(label + ' ' + bar + ' ' + timeStr + ' ' + annotation);
    }

    return lines.join('\n');
}

// Generate section for one category
function generateCategorySection(category, benchGroups) {
    const meta = CATEGORY_META[category] || {};
    let section = `## ${category}\n\n`;
    if (meta.description) section += `> ${meta.description}\n\n`;

    for (const [benchName, entries] of Object.entries(benchGroups)) {
        section += generateTable(entries, benchName);
        section += '\n**Comparison** (lower is better):\n```\n';
        section += generateBarChart(entries);
        section += '\n```\n\n';
    }

    return section;
}

// Generate full report
function generateReport(results) {
    const sysInfo = getSystemInfo();

    let report = '# HTTP Parser Benchmark Results\n\n';
    report += '## System Information\n\n';
    report += '| Item | Value |\n';
    report += '|------|-------|\n';
    report += `| OS | ${sysInfo.platform} ${sysInfo.osVersion} |\n`;
    report += `| Architecture | ${sysInfo.arch} |\n`;
    report += `| CPU | ${sysInfo.cpu} (${sysInfo.cpuCores} cores) |\n`;
    report += `| Memory | ${sysInfo.memory} |\n`;
    report += `| Compiler | ${sysInfo.compiler} |\n`;
    report += `| Build Flags | -O2 |\n`;
    report += '\n---\n\n';

    const groups = groupResults(results);

    // Emit categories in defined order, then any remaining
    const emitted = new Set();
    for (const cat of CATEGORY_ORDER) {
        if (groups[cat]) {
            report += generateCategorySection(cat, groups[cat]);
            report += '---\n\n';
            emitted.add(cat);
        }
    }
    for (const cat of Object.keys(groups)) {
        if (!emitted.has(cat)) {
            report += generateCategorySection(cat, groups[cat]);
            report += '---\n\n';
        }
    }

    report += '_Generated by report.js_\n';
    return report;
}

// Main
function main() {
    const args = process.argv.slice(2);
    if (args.length < 1) {
        console.error('Usage: node report.js <file.jsonl> [...]');
        console.error('  node report.js results/req_*.jsonl   # all request results (shell glob)');
        console.error('  node report.js results/rsp_*.jsonl   # all response results');
        console.error('  node report.js results/req_hwire.jsonl  # hwire only');
        process.exit(1);
    }

    const files = args.filter(f => {
        if (!fs.existsSync(f)) {
            console.warn(`Warning: file not found: ${f}`);
            return false;
        }
        return true;
    });

    const results = parseJsonlFiles(files);
    if (results.length === 0) {
        console.log('# HTTP Parser Benchmark Results\n\n_No benchmark results found_\n');
        process.exit(0);
    }

    console.log(generateReport(results));
}

main();

File: bench/report.py
#!/usr/bin/env python3
"""Compare benchmark results written by `make run` into results/.

Each result file is named <parser>-<variant>.txt and contains one line
per benchmark:

    <fixture>/<bytes> <samples> <iterations> <mean-ns> <stddev-ns>

The report is organized per variant: every variant found in the result
files gets its own set of tables (time, throughput) containing the
parsers measured under that build configuration, with ratios relative to
hwire under the same variant. Files without a known variant suffix form
a "default" group.
"""

import sys
from pathlib import Path

RESULTS_DIR = Path(__file__).resolve().parent / "results"
PLATFORM_FILE = RESULTS_DIR / "platform.txt"

CANONICAL_VARIANTS = ["nosimd", "sse2", "neon", "sse42", "avx2"]

def load_results(directory: Path) -> dict[str, dict[str, dict]]:
    """row name -> fixture -> {bytes, samples, iterations, mean, stddev}."""
    data: dict[str, dict[str, dict]] = {}
    for path in sorted(directory.glob("*.txt")):
        if path.name == PLATFORM_FILE.name:
            continue  # environment info, not benchmark results
        row: dict[str, dict] = {}
        for line in path.read_text().splitlines():
            fields = line.split()
            if not fields:
                continue
            name, _, size = fields[0].rpartition("/")
            entry = {"bytes": int(size) if size else None}
            try:
                values = [float(v) for v in fields[1:]]
                if len(values) == 1:
                    entry.update(samples=None, iterations=None,
                                 mean=values[0], stddev=0.0)
                else:
                    (entry["samples"], entry["iterations"], entry["mean"],
                     entry["stddev"]) = values
            except ValueError:
                sys.exit(f"{path.name}: unparsable line: {line!r}")
            row[name] = entry
        if row:
            data[path.stem] = row
    return data


def row_sort_key(name: str) -> tuple:
    # hwire first (it is the library under test), others alphabetically
    return (0 if name == "hwire" else 1, name)


def split_variant(name: str) -> tuple[str, str]:
    """'picohttpparser-sse42' -> ('picohttpparser', 'sse42')."""
    base, _, variant = name.rpartition("-")
    if base and variant in CANONICAL_VARIANTS:
        return base, variant
    return name, "default"


def human_throughput(nbytes: float, ns: float) -> str:
    bps = nbytes / (ns * 1e-9)
    for unit, scale in (("GB/s", 1e9), ("MB/s", 1e6), ("kB/s", 1e3)):
        if bps >= scale:
            return f"{bps / scale:.2f} {unit}"
    return f"{bps:.2f} B/s"


def render(title: str, header: list[str], rows: list[list[str]]) -> None:
    widths = [max(len(header[i]), *(len(r[i]) for r in rows))
              for i in range(len(header))]
    line = "  ".join(h.ljust(w) for h, w in zip(header, widths))
    print(title)
    print(line)
    print("-" * len(line))
    for row in rows:
        print("  ".join(c.ljust(w) for c, w in zip(row, widths)))
    print()


def print_platform() -> None:
    """Show the environment that produced the results (collected at
    measurement time by platform.sh), if it is available."""
    if not PLATFORM_FILE.is_file():
        return
    print("platform")
    print("--------")
    for line in PLATFORM_FILE.read_text().splitlines():
        print(f"  {line}")
    print()


def main() -> None:
    if not RESULTS_DIR.is_dir():
        sys.exit(f"no results directory at {RESULTS_DIR} (run `make run` first)")

    data = load_results(RESULTS_DIR)
    if not data:
        sys.exit(f"no result files in {RESULTS_DIR} (run `make run` first)")

    print_platform()

    # Group rows per variant; order variants canonically, "default" first.
    groups: dict[str, list[str]] = {}
    for name in data:
        groups.setdefault(split_variant(name)[1], []).append(name)
    variant_order = ([v for v in CANONICAL_VARIANTS if v in groups]
                     + [v for v in sorted(groups) if v not in CANONICAL_VARIANTS
                        and v != "default"])
    if "default" in groups:
        variant_order = ["default"] + variant_order
    for rows in groups.values():
        rows.sort(key=row_sort_key)

    # Fixture byte sizes (per fixture, from whichever row has them)
    sizes = {}
    for row in data:
        for fixture, entry in data[row].items():
            if entry["bytes"]:
                sizes.setdefault(fixture, entry["bytes"])

    for variant in variant_order:
        rows = groups[variant]
        baseline = next((r for r in rows if r == f"hwire-{variant}"), rows[0])

        fixtures: list[str] = []
        for row in rows:
            for fixture in data[row]:
                if fixture not in fixtures:
                    fixtures.append(fixture)

        print(f"===== variant: {variant} "
              f"(base = {baseline}) " + "=" * max(0, 40 - len(variant)))
        print()

        # time per message
        out = []
        for row in rows:
            cells = [row]
            for fixture in fixtures:
                entry = data[row].get(fixture)
                if entry is None:
                    cells.append("-")
                    continue
                cell = f"{entry['mean']:.1f} ±{entry['stddev']:.1f}"
                if len(rows) > 1:
                    base = data[baseline].get(fixture)
                    if base:
                        cell += (f" (x{entry['mean'] / base['mean']:.2f})"
                                 if row != baseline else " (base)")
                cells.append(cell)
            out.append(cells)
        render("time per message, ns (lower is better)",
               ["parser"] + [f"{f} ({sizes[f]}B)" if f in sizes else f
                             for f in fixtures], out)

        # throughput
        out = []
        for row in rows:
            cells = [row]
            for fixture in fixtures:
                entry = data[row].get(fixture)
                if entry is None or not entry["bytes"]:
                    cells.append("-")
                else:
                    cells.append(human_throughput(entry["bytes"],
                                                  entry["mean"]))
            out.append(cells)
        render("throughput (higher is better)", ["parser"] + fixtures, out)

    # measurement note (identical for the shared driver)
    first = next(iter(next(iter(data.values())).values()))
    if first.get("samples"):
        print(f"measurement: {int(first['samples'])} samples x "
              f"{int(first['iterations'])} iterations per benchmark")
    print("(xN.NN = N.NN times the base parser's time, same variant)")


if __name__ == "__main__":
    main()
